File: app/src/main/java/io/mgba/Activities/MainActivity.java
package io.mgba.Activities;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.TabLayout;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.view.MenuItem;

import com.arlib.floatingsearchview.FloatingSearchView;
import com.arlib.floatingsearchview.suggestions.model.SearchSuggestion;
import com.flipboard.bottomsheet.BottomSheetLayout;
import com.mikepenz.aboutlibraries.LibsBuilder;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.mgba.Adapters.TabViewPager;
import io.mgba.Data.DTOs.Game;
import io.mgba.Model.Interfaces.ILibrary;
import io.mgba.Model.Library;
import io.mgba.R;
import io.mgba.Views.BottomSheetView;
import io.mgba.Views.Interfaces.IBottomSheetView;
import io.mgba.mgba;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class MainActivity extends AppCompatActivity implements TabLayout.OnTabSelectedListener, FloatingSearchView.OnMenuItemClickListener, FloatingSearchView.OnQueryChangeListener, FloatingSearchView.OnSearchListener, io.mgba.Activities.Interfaces.ILibrary, FloatingSearchView.OnFocusChangeListener {

    private static final int DEFAULT_PANEL = 1;
    private static final int SETTINGS_CODE = 738;

    @BindView(R.id.floating_search_view) FloatingSearchView mToolbar;
    @BindView(R.id.pager) ViewPager mViewPager;
    @BindView(R.id.bottomsheet) BottomSheetLayout mSheetDialog;
    @BindView(R.id.tabLayout) TabLayout mTabLayout;

    private ILibrary libraryService;

    private IBottomSheetView bottomSheetController;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_library);

        ButterKnife.bind(this);

        libraryService = new Library((mgba)getApplication());

        prepareToolbar();
        prepareViewPager();
    }

    private void prepareToolbar() {
        mToolbar.setOnMenuItemClickListener(this);
        mToolbar.setOnSearchListener(this);
        mToolbar.setOnQueryChangeListener(this);
        mToolbar.setOnFocusChangeListener(this);
    }

    private void prepareViewPager() {
        mTabLayout.addTab(mTabLayout.newTab().setText(getString(R.string.Favorites)));
        mTabLayout.addTab(mTabLayout.newTab().setText(getString(R.string.GBA)));
        mTabLayout.addTab(mTabLayout.newTab().setText(getString(R.string.GBC)));
        mTabLayout.setTabGravity(TabLayout.GRAVITY_FILL);

        TabViewPager adapter = new TabViewPager(getSupportFragmentManager());

        mViewPager.setAdapter(adapter);
        mViewPager.setCurrentItem(DEFAULT_PANEL);

        mViewPager.addOnPageChangeListener(new TabLayout.TabLayoutOnPageChangeListener(mTabLayout));
        mTabLayout.addOnTabSelectedListener(this);
    }

    @Override
    public void onTabSelected(TabLayout.Tab tab) {
        mViewPager.setCurrentItem(tab.getPosition());
    }

    @Override
    public void onTabUnselected(TabLayout.Tab tab) {

    }

    @Override
    public void onTabReselected(TabLayout.Tab tab) {

    }

    @Override
    public void onActionMenuItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.action_about)
        {
            new LibsBuilder()
                    .withActivityTheme(R.style.AboutTheme)
                    .withAboutIconShown(true)
                    .withAboutVersionShown(true)
                    .withAboutDescription(getString(R.string.About_description))
                    .start(this);
        }

        if (item.getItemId() == R.id.action_settings)
        {
            Intent settings = new Intent(this, SettingsCategoriesActivity.class);
            startActivityForResult(settings, SETTINGS_CODE);
        }
    }

    @Override
    public void showBottomSheet(Game game) {
        if(bottomSheetController == null)
            bottomSheetController = new BottomSheetView(getApplicationContext());

        mSheetDialog.showWithSheetView(bottomSheetController.getView(mSheetDialog, game));
    }

    @Override
    public ILibrary getLibraryService() {
        return libraryService;
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent intent) {
        if (requestCode == SETTINGS_CODE && resultCode == Activity.RESULT_OK) {

        }
    }

    @Override
    public void onSuggestionClicked(SearchSuggestion searchSuggestion) {

    }

    @Override
    public void onSearchAction(String currentQuery) {
        //Not implemented
    }

    @Override
    public void onSearchTextChanged(String oldQuery, String newQuery) {

        if (!oldQuery.equals("") && newQuery.equals("")) {
            mToolbar.clearSuggestions();
        } else {
            mToolbar.showProgress();
            libraryService
                    .query(newQuery)
                    .subscribeOn(Schedulers.computation())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(games -> {
                        mToolbar.swapSuggestions(games);
                        mToolbar.hideProgress();
                    });
        }
    }

    @Override
    public void onFocus() {
        onSearchTextChanged("", mToolbar.getQuery());
    }

    @Override
    public void onFocusCleared() {

    }
}

File: app/src/main/java/io/mgba/main/MainActivity.kt
package io.mgba.main

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.os.Parcelable
import android.view.MenuItem
import com.arlib.floatingsearchview.FloatingSearchView
import com.arlib.floatingsearchview.suggestions.model.SearchSuggestion
import com.google.android.material.tabs.TabLayout
import com.mikepenz.aboutlibraries.LibsBuilder
import androidx.appcompat.app.AppCompatActivity
import io.mgba.main.adapters.TabViewPager
import io.mgba.data.interfaces.ILibrary
import io.mgba.main.MainViewModel.Companion.DEFAULT_PANEL
import io.mgba.R
import io.mgba.base.BaseActivity
import io.mgba.mgba
import io.mgba.ui.activities.interfaces.IMainView
import kotlinx.android.synthetic.main.activity_library.*

class MainActivity(override val libraryService: ILibrary) : BaseActivity<MainViewModel>(), TabLayout.OnTabSelectedListener, io.mgba.ui.activities.interfaces.ILibrary, FloatingSearchView.OnMenuItemClickListener, FloatingSearchView.OnQueryChangeListener, FloatingSearchView.OnSearchListener, IMainView {

    override fun getLayout(): Int = R.layout.activity_library
    override fun getViewModel(): Class<MainViewModel> = MainViewModel::class.java

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        prepareToolbar()
        prepareViewPager()

        if (savedInstanceState != null) {
            floatingSearchView.onRestoreInstanceState(savedInstanceState.getParcelable<Parcelable>(MAIN_TOOLBAR_STATE))
            pager.onRestoreInstanceState(savedInstanceState.getParcelable<Parcelable>(MAIN_VIEWPAGE_STATE))
        }
    }

    private fun prepareToolbar() {
        floatingSearchView.setOnMenuItemClickListener(this)
        floatingSearchView.setOnSearchListener(this)
        floatingSearchView.setOnQueryChangeListener(this)
    }

    private fun prepareViewPager() {
        tabLayout.addTab(tabLayout.newTab().setText(getString(R.string.Favorites)))
        tabLayout.addTab(tabLayout.newTab().setText(getString(R.string.GBA)))
        tabLayout.addTab(tabLayout.newTab().setText(getString(R.string.GBC)))
        tabLayout.tabGravity = TabLayout.GRAVITY_FILL

        val adapter = TabViewPager(supportFragmentManager)

        pager.adapter = adapter
        pager.currentItem = DEFAULT_PANEL

        pager.addOnPageChangeListener(TabLayout.TabLayoutOnPageChangeListener(tabLayout))
        tabLayout.addOnTabSelectedListener(this)
    }

    override fun onTabSelected(tab: TabLayout.Tab) {
        pager.currentItem = tab.position
    }

    override fun onTabReselected(tab: TabLayout.Tab) {
        pager.currentItem = tab.position
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, intent: Intent?) {
        if(intent != null) {} //controller.onActivityResult(requestCode, resultCode, intent)
    }

    override fun onTabUnselected(p0: TabLayout.Tab?) {}
    override fun onSuggestionClicked(searchSuggestion: SearchSuggestion?) {}
    override fun onSearchAction(currentQuery: String?) {}

    override fun onSearchTextChanged(oldQuery: String, newQuery: String) {
        //controller.onSearchTextChanged(oldQuery, newQuery)
    }

    override fun onActionMenuItemSelected(item: MenuItem) {
        //controller.onMenuItemSelected(item)
    }

    override fun clearSuggestions() {
        floatingSearchView.clearSuggestions()
    }

    override fun showSuggestions(list: List<SearchSuggestion>) {
        floatingSearchView.swapSuggestions(list)
        floatingSearchView.hideProgress()
    }

    override fun startAboutPanel(aboutPanel: LibsBuilder) {
        aboutPanel.start(this)
    }

    override fun startActivityForResult(activity: Class<out AppCompatActivity>, code: Int) {
        val it = Intent(this, activity)
        startActivityForResult(it, code)
    }

    override fun showProgress() {
        floatingSearchView.showProgress()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelable(MAIN_TOOLBAR_STATE, floatingSearchView.onSaveInstanceState())
        outState.putParcelable(MAIN_VIEWPAGE_STATE, pager.onSaveInstanceState())
    }


    companion object {
        fun start() {
            val it = Intent(mgba.context, MainActivity::class.java)
            mgba.context.startActivity(it)
        }

        val MAIN_TOOLBAR_STATE = "main_toolbar_state"
        val MAIN_VIEWPAGE_STATE = "main_viewpage_state"
    }
}
